src/st/st0/3D8F0.c

```diff
@@ -1,5 +1,266 @@
+        if (dr_env == NULL) {
+            SetStep(7);
+            return;
+        }
+        prim->type = PRIM_ENV;
+        prim->priority = 0x6A;
+        prim->drawMode = DRAW_UNK_800;
+        self->step++;
+        return;
+    case 4:
+        prim = self->ext.prim;
+        prim = prim->next;
+        for (i = 0; i < 2; i++) {
+            dr_env = LOW(prim->r1);
+            dr_env->tag = 0;
+            prim->type = PRIM_GT4;
+            prim = prim->next;
+        }
+        prim = self->ext.prim;
+        for (i = 0; i < 15; i++) {
+            for (j = 0; j < 15; j++) {
+                xVar = j * 17;
+                yVar = i * 16;
+                prim->tpage = 0x10C;
+                prim->u0 = prim->u2 = xVar;
+                prim->u1 = prim->u3 = xVar + 17;
+                prim->v0 = prim->v1 = yVar;
+                prim->v2 = prim->v3 = yVar + 16;
+                prim->x0 = prim->x2 = xVar;
+                prim->x1 = prim->x3 = xVar + 17;
+                prim->y0 = prim->y1 = yVar;
+                prim->y2 = prim->y3 = yVar + 16;
+                prim->r0 = 0x40;
+                prim->g0 = prim->b0 = 0x30;
+
+                LOW(prim->r1) = LOW(prim->r0);
+                LOW(prim->r2) = LOW(prim->r0);
+                LOW(prim->r3) = LOW(prim->r0);
+                prim->priority = 0x6C;
+                prim->drawMode = DRAW_DEFAULT;
+                prim = prim->next;
+            }
+        }
+        self->step++;
+        return;
+    case 5:
+        var_s3 = &D_801C06B0;
+        g_Tilemap.flags &= 0xFFFE;
+        g_BgLayers[0].flags &= 0xFFFE;
+        prim = self->ext.prim;
+        for (i = 0; i < 15 * 15; i++, var_s3++) {
+            xVar = prim->x0 + 8;
+            yVar = prim->y0 + 8;
+            xVar -= 0x80;
+            yVar -= 0x80;
+            var_s3->unkA = xVar;
+            var_s3->unkE = yVar;
+            var_s3->unk10 = 0x80;
+            angle = ratan2(yVar, xVar);
+            var_s7 = (xVar * xVar) + (yVar * yVar);
+            xVar = rsin(angle) * 0x24;
+            yVar = rcos(angle) * 0x24;
+            var_s7 = SquareRoot0(var_s7);
+            var_s8 = 0xC0 - var_s7;
+            xVar = xVar * var_s8 / 0xC0;
+            yVar = yVar * var_s8 / 0xC0;
+            xVar >>= 0xC;
+            yVar >>= 0xC;
+            var_s3->unk0.x = xVar;
+            var_s3->unk0.y = yVar;
+            var_s3->unk14 = ((var_s7 * 3) / 2) - (Random() & 0x1F);
+            if (var_s3->unk14 < 0) {
+                var_s3->unk14 = 0;
+            }
+            var_s3->unk4 = 0;
+            var_s3->unk6 = 0;
+            prim = prim->next;
+        }
+        self->ext.GS_Props.pickupFlag = 1;
+        self->step++;
+        return;
+    case 6:
+        SetGeomScreen(0x80);
+        SetGeomOffset(0x80, 0x80);
+        sp38 = (MATRIX*)SPAD(0);
+        sp40 = (SVECTOR*)SPAD(8);
+        sp3c = (VECTOR*)SPAD(10);
+        sp34 = (SVECTOR*)SPAD(14);
```

Comment:
Not certain how/if these SPAD calls need to be replaced for PC, would appreciate suggestions.

Reply:
```
typedef struct {
    short m[3][3]; // 3 x 3 matrix coefficient value
    long t[3];     // Parallel transfer volume
} MATRIX;          // size = 0x1E
```
It looks like these overlap and it's all one matrix? You need something like
#ifdef VERSION_PC
MATRIX *sp38;
MATRIX sp38_;
sp38 = &sp38_;
#endif

Reply:
They don't overlap. But thanks, I think I've made an implementation that should work.

The matrix is 30 bytes long, and the sp40 vector starts 32 bytes deep in the scratchpad, leaving 2 bytes unused. Then the 3 vectors are consecutive.